modules/navier_stokes/src/fv/bcs/NSFVFunctionBC.C

```diff
@@ -0,0 +1,143 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NSFVFunctionBC.h"
+
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+
+#include "MooseVariableFieldBase.h"
+#include "SystemBase.h"
+#include "ADReal.h"    // Moose::derivInsert
+#include "MooseMesh.h" // FaceInfo methods
+#include "FVDirichletBC.h"
+
+#include "libmesh/dof_map.h"
+#include "libmesh/elem.h"
+#include "libmesh/numeric_vector.h"
+#include "libmesh/vector_value.h"
+
+registerMooseObject("NavierStokesApp", NSFVFunctionBC);
+
+InputParameters
+NSFVFunctionBC::validParams()
+{
+  InputParameters params = FVMatAdvectionFunctionBC::validParams();
+  params += NSFVBase::validParams();
+
+  params.addRequiredParam<FunctionName>("pressure_exact_solution",
+                                        "The function describing the pressure exact solution.");
+  return params;
+}
+
+NSFVFunctionBC::NSFVFunctionBC(const InputParameters & params)
+  : FVMatAdvectionFunctionBC(params),
+    NSFVBase(params),
+    _pressure_exact_solution(getFunction("pressure_exact_solution"))
+{
+}
+
+void
+NSFVFunctionBC::interpolate(Moose::FV::InterpMethod m,
+                            ADRealVectorValue & v_face,
+                            const ADRealVectorValue & elem_v,
+                            const RealVectorValue & ghost_v)
+{
+  Moose::FV::interpolate(
+      Moose::FV::InterpMethod::Average, v_face, elem_v, ghost_v, *_face_info, true);
+
+  if (m == Moose::FV::InterpMethod::RhieChow)
+  {
+    auto tup = Moose::FV::determineElemOneAndTwo(*_face_info, *_p_var);
+    const Elem * const elem = std::get<0>(tup);
+    const bool elem_is_elem = std::get<2>(tup);
+    mooseAssert(elem_is_elem ? elem == &_face_info->elem() : elem == _face_info->neighborPtr(),
+                "elem_is_elem is incorrect");
+
+    const Point & elem_centroid =
+        elem_is_elem ? _face_info->elemCentroid() : _face_info->neighborCentroid();
+    const Real elem_volume = elem_is_elem ? _face_info->elemVolume() : _face_info->neighborVolume();
+
+    // Get pressure gradient for the elem
+    const VectorValue<ADReal> & grad_p_elem = _p_var->adGradSln(elem);
+
+    // Get pressure gradient for the ghost
+    RealVectorValue grad_p_ghost =
+        _pressure_exact_solution.gradient(_t, 2. * _face_info->faceCentroid() - elem_centroid);
+
+    // Uncorrected face pressure gradient
+    auto unc_grad_p = (grad_p_elem + grad_p_ghost) / 2.;
+
+    // Now perform correction
+
+    // Get pressure value for the elem
+    ADReal p_elem = _p_var->getElemValue(elem);
+
```

Comment:
`elem` here may not be the `FaceInfo` `elem`. It could be the `FaceInfo neighbor` depending on subdomain restriction of variables. We would have to do some boolean logic to figure out whether to use the reference coming from `adCoupledValue` or `adCoupledNeighborValue`. What I like about calling these `getElemValue` APIs is that it is abundantly clear what I am doing, and I can trust the result that I am getting. This kind of goes along the lines (I think) of what @rwcarlsen wishes our material system was like: call a function to get the value of something where and when you want it. I have a hard time sometimes tracking our system of references and reinits behind the scenes. Things are fine when you are only on a single element, but at least for me introducing even one other element (e.g. the neighbor) makes the behind-the-scenese-reinit harder to grok (understand).

The (perhaps) unfortunate bit here is that I am returning an `ADReal` from `getElemValue` as opposed to a `const ADReal &`. Maybe I should make a cache for these like I do for our gradients...